calc/vxlan_resolver.go

```diff
@@ -0,0 +1,476 @@
+		return false
+	}
+	return true
+}
+
+// kickPendingRoutes loops through the provided routes to see if there are any which are now programmable,
+// and will send any that are.
+func (c *VXLANResolver) kickPendingRoutes(pendingRouteUpdates set.Set) {
+	pendingRouteUpdates.Iter(func(item interface{}) error {
+		r := item.(vxlanRoute)
+		logCxt := logrus.WithField("route", r)
+		if !c.routeReady(r) {
+			return nil
+		}
+		logCxt.Debug("Sending VXLAN route update")
+		c.callbacks.OnRouteUpdate(&proto.RouteUpdate{
+			Type: proto.RouteType_VXLAN,
+			Node: r.node,
+			Dst:  r.dst.String(),
+			Gw:   c.determineGatewayForRoute(r),
+		})
+		return nil
+	})
+}
+
+// withdrawRoute will send a *proto.RouteRemove for the given route.
+func (c *VXLANResolver) withdrawRoute(r vxlanRoute) {
+	logrus.WithField("route", r).Debug("Sending VXLAN route remove")
+	c.callbacks.OnRouteRemove(r.dst.String())
+}
+
+func (c *VXLANResolver) sendVTEPUpdate(node string) bool {
+	logCxt := logrus.WithField("node", node)
+	tunlAddr, ok := c.nodeNameToVXLANTunnelAddr[node]
+	if !ok {
+		logCxt.Debug("Missing vxlan tunnel address for node, cannot send VTEP")
+		return false
+	}
+	parentDeviceIP, ok := c.nodeNameToIPAddr[node]
+	if !ok {
+		logCxt.Debug("Missing IP for node, cannot send VTEP")
+		return false
+	}
+
+	logCxt.Debug("Sending VTEP to dataplane")
+	vtep := &proto.VXLANTunnelEndpointUpdate{
+		Node:           node,
+		ParentDeviceIp: parentDeviceIP,
+		Mac:            c.vtepMACForHost(node),
+		Ipv4Addr:       tunlAddr,
+	}
+	c.callbacks.OnVTEPUpdate(vtep)
+	return true
+}
+
+func (c *VXLANResolver) sendVTEPRemove(node string) {
+	logrus.WithField("node", node).Debug("Withdrawing VTEP from dataplane")
+	c.callbacks.OnVTEPRemove(node)
+}
+
+// routeWithinVXLANPool checks if the provided route is within a configured IP pool with
+// VXLAN enabled.
+func (c *VXLANResolver) routeWithinVXLANPool(r vxlanRoute) bool {
+	for _, pool := range c.vxlanPools {
+		if c.containsRoute(pool, r) {
+			return true
+		}
+	}
+	return false
+}
+
+// routesFromBlock returns a list of routes which should exist based on the provided
+// allocation block. Right now, we only support strict affinity so this will always
+// be the block's CIDR.
+func (c *VXLANResolver) routesFromBlock(blockKey string, b *model.AllocationBlock) map[string]vxlanRoute {
+	if b.Host() == c.hostname {
+		logrus.Debug("Skipping VXLAN routes for local node")
+		return nil
+	}
+
+	r := vxlanRoute{
+		dst:  ip.CIDRFromCalicoNet(b.CIDR),
+		node: b.Host(),
+	}
+	return map[string]vxlanRoute{r.Key(): r}
+}
+
+// determineGatewayForRoute determines which gateway to use for this route. For VXLAN routes, the
+// gateway is the remote node's IPv4VXLANTunnelAddr. If we don't know the remote node's tunnel
+// address, this function will return an empty string.
+func (c *VXLANResolver) determineGatewayForRoute(r vxlanRoute) string {
+	if ip, ok := c.nodeNameToVXLANTunnelAddr[r.node]; ok {
```

Comment:
You could just do `return c.nodeNameToVXLANTunnelAddr[r.node]` since accessing a missing map entry returns the zero value for the type.